### PR TITLE
Discrete disclosure risk metric helper functions

### PR DESCRIPTION
**Pull request by jhseeman**

### CODE REVIEW

File: R/disc_dd_helper.R [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  
+  # determine if using inferred keys or user-specified keys
+  if (is.null(keys)) {
+    
+    keys <- names(eval_data$conf_data)
+    
+  } else {
+    
+    stopifnot(all(keys %in% names(eval_data$conf_data)))
+    keys <- keys
```
Comment (awunderground) [inline R/disc_dd_helper.R:49]: Why is this line necessary?

File: R/disc_dd_helper.R [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  } 
+  
+  # check all confidential data keys present in synthetic data 
+  if (eval_data$n_rep > 1) {
+   
+    for (ix in seq(eval_data$n_rep)) {
+      
+      stopifnot(all(keys %in% names(eval_data$synth_data[[ix]])))
+      
+    }
```
Comment (awunderground) [inline R/disc_dd_helper.R:60]: There is probably a clever way to replace this with `walk()` but it isn't necessary!

File: R/disc_dd_helper.R [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        )
+        
+      }
+      
+    }
+    
+  }
+  
+  # return 0 if all checks passed
+  return(0)
```
Comment (awunderground) [inline R/disc_dd_helper.R:127]: `0` is a surprising value to return for a validation. Did you think about `TRUE`/`FALSE`?

File: R/disc_dd_helper.R [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+      
+      col_min_candidates <- c(
+        col_min_candidates, 
+        min(dplyr::pull(eval_data$synth_data, col), na.rm = TRUE)
+      )
+      
+      col_max_candidates <- c(
+        col_max_candidates,
+        max(dplyr::pull(eval_data$synth_data, col), na.rm = TRUE)
+      )
```
Comment (awunderground) [inline R/disc_dd_helper.R:181]: It probably doesn't matter in this context, but growing vectors in R is very slow.

File: R/disc_dd_helper.R [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        col_min_candidates, 
+        min(dplyr::pull(eval_data$holdout_data, col), na.rm = TRUE)
+      )
+      
+      col_max_candidates <- c(
+        col_max_candidates,
+        max(dplyr::pull(eval_data$holdout_data, col), na.rm = TRUE)
+      )
+      
+    }
```
Comment (awunderground) [inline R/disc_dd_helper.R:219]: This entire section would benefit from better commenting. Why are you doing each of these steps?

File: R/disc_dd_helper.R [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    
+    # define mutate function 
+    discretize_col <- function(data, col_name, breaks) {
+      
+      # apply non-NA breaks
+      non_na_cut <- base::cut(
+        dplyr::pull(data, col_name), 
+        breaks
+      )
```
Comment (awunderground) [inline R/disc_dd_helper.R:251]: Please include argument names.

Comment (awunderground) [inline R/disc_dd_helper.R:251]: `cut()` doesn't include the minimum values by default and instead introduces `NA`. Consider the following example using the `cars` data set:

```r
> orig_ed <- eval_data(
+   conf_data = cars, 
+   synth_data = cars
+ )
> 
> disc_ed1 <- prep_discrete_eval_data(
+   orig_ed,
+   col_map = list(
+     "speed" = list("k" = 2), 
+     "dist" = list("k" = 2)
+   )
+ )
> 
> head(disc_ed1[["conf_data"]])
     speed   dist
1     <NA>   <NA>
2     <NA> (2,61]
3 (4,14.5] (2,61]
4 (4,14.5] (2,61]
5 (4,14.5] (2,61]
6 (4,14.5] (2,61]
```

Please turn this into a test to ensure the function doesn't introduce additional `NA` values into the confidential data.

Comment (awunderground) [reply R/disc_dd_helper.R:251]: I *think* adding `include.lowest = TRUE` helps. 

```r
> x <- 1:10
> 
> cut(x = x, breaks = c(1, 5, 10))
 [1] <NA>   (1,5]  (1,5]  (1,5]  (1,5]  (5,10] (5,10] (5,10] (5,10] (5,10]
Levels: (1,5] (5,10]
> 
> cut(x = x, breaks = c(1, 5, 10), include.lowest = TRUE)
 [1] [1,5]  [1,5]  [1,5]  [1,5]  [1,5]  (5,10] (5,10] (5,10] (5,10] (5,10]
Levels: [1,5] (5,10]
```

**Review by awunderground — CHANGES_REQUESTED** [review]
The issue with `cut()` is major and needs a fix. Everything else is pretty minor.

File: R/disc_dd_helper.R (hunk 1 above)
Comment (jhseeman) [reply R/disc_dd_helper.R:49]: it is not! good catch

File: R/disc_dd_helper.R (hunk 3 above)
Comment (jhseeman) [reply R/disc_dd_helper.R:127]: I was using the linux exit code convention https://www.agileconnection.com/article/overview-linux-exit-codes but open to alternatives if you feel strongly another way

File: R/disc_dd_helper.R (hunk 4 above)
Comment (jhseeman) [reply R/disc_dd_helper.R:181]: yep, assuming syntheval users are not preparing egregiously wide tables for evaluation, pre-allocating the space for the vector would make a negligible difference. Can revisit if we choose to make this scalable (ex: incorporating Spark SQL or similar tools)

File: R/disc_dd_helper.R (hunk 5 above)
Comment (jhseeman) [reply R/disc_dd_helper.R:219]: flag for later: need to update this to consider `na.rm` logic!

**Review by awunderground — APPROVED** [review]
Good work.

**Pull request merged**